xmind/tests/test_workbook_saver.py

```diff
@@ -0,0 +1,135 @@
+            _utils_join_path.return_value = 'joined path'
+            with patch('xmind.utils.temp_dir') as _utils_temp_dir:
+                _utils_temp_dir.return_value = 'temp dir'
+                _m = mock_open()  # to test with statement
+                with patch('codecs.open', _m, create=True):
+                    with patch.object(self._workbook, 'output') as _output:
+                        _el = self._test_el._get_content()
+
+        _utils_temp_dir.assert_called_once_with()
+        _utils_join_path.assert_called_once_with('temp dir', CONTENT_XML)
+        _m.assert_called_once_with('joined path', 'w', encoding="utf-8")
+        _output.assert_called_once_with(_m())
+        self.assertEqual(_el, 'joined path')
+
+    def test_save_with_no_path_no_get_path(self):
+        with patch.object(self._workbook, 'get_path') as _get_path:
+            _get_path.return_value = None
+            _test_el = WorkbookSaver(self._workbook)
+            with self.assertRaises(Exception) as _ex:
+                _test_el.save()
+
+        self.assertTrue(_ex.exception.args[0].find("Please specify a filename for the XMind file") != -1)
+        _get_path.assert_called_once_with()
+
+    def test_save_is_path(self):
+        with patch('xmind.utils.get_abs_path') as _get_abs_path:
+            _get_abs_path.side_effect = Exception('stop point')
+
+            with self.assertRaises(Exception) as _ex:
+                self._test_el.save('path')
+
+        self.assertTrue(_ex.exception.args[0].find("stop point") == 0)
+        _get_abs_path.assert_called_once_with('path')
+
+    def test_save_no_path_is_get_path(self):
+        with patch('xmind.utils.get_abs_path') as _get_abs_path:
+            _get_abs_path.side_effect = Exception('stop point')
+
+            with patch.object(self._workbook, 'get_path') as _get_path:
+                _get_path.return_value = 'get path'
+                with self.assertRaises(Exception) as _ex:
+                    self._test_el.save()
+
+        self.assertTrue(_ex.exception.args[0].find("stop point") == 0)
+        _get_abs_path.assert_called_once_with('get path')
+        _get_path.assert_called_once_with()
+
+    def test_save_if_ext_is_not_equal_xmind_ext(self):
+        with patch('xmind.utils.get_abs_path') as _get_abs_path:
+            _get_abs_path.return_value = 'abs path'
+            with patch('xmind.utils.split_ext') as _split_ext:
+                _split_ext.return_value = 'name', 'ext'
+                with self.assertRaises(Exception) as _ex:
+                    self._test_el.save('path')
+
+        self.assertTrue(_ex.exception.args[0].find("XMind filenames require a '%s' extension" % XMIND_EXT) != -1)
+        _get_abs_path.assert_called_once_with('path')
+        _split_ext.assert_called_once_with('abs path')
+
+    def test_save_if_ext_is_equal_xmind_ext(self):
+        with patch('xmind.utils.get_abs_path') as _get_abs_path:
```

Comment:
More than 2 levels of nesting is hard to follow

Reply:
Done